pkg/operator2/datasync/validation.go

```diff
@@ -0,0 +1,136 @@
+package datasync
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	corelistersv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/keyutil"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+var validators = map[string]func(data []byte) []error{
+	corev1.TLSCertKey:       validateClientCert,
+	corev1.TLSPrivateKeyKey: validatePrivateKey,
+
+	corev1.ServiceAccountRootCAKey: validateCACerts,
+	configv1.ClientSecretKey:       noValidation,
+	configv1.HTPasswdDataKey:       noValidation,
+	configv1.BindPasswordKey:       noValidation,
+}
+
+func noValidation(_ []byte) []error { return []error{} }
+
+func validateSecret(secretsLister corelistersv1.SecretLister, src sourceData) []error {
+	s, err := secretsLister.Secrets("openshift-config").Get(src.Src)
+	if err != nil {
+		return []error{err}
+	}
+
+	data, exists := s.Data[src.Key]
+	if !exists {
+		return []error{fmt.Errorf("missing required key: %q", src.Key)}
+	}
+
+	return validators[src.Key](data)
+}
+
+func validateConfigMap(cmLister corelistersv1.ConfigMapLister, src sourceData) []error {
+	cm, err := cmLister.ConfigMaps("openshift-config").Get(src.Src)
+	if err != nil {
+		return []error{err}
+	}
+
+	data, exists := cm.Data[src.Key]
+	if !exists {
+		return []error{fmt.Errorf("missing required key: %q", src.Key)}
+	}
+
+	return validators[src.Key]([]byte(data))
+}
+
+func validateClientCert(pem []byte) []error {
+	errs := []error{}
+
+	certs, certErrs := parseCerts(pem)
+	errs = append(errs, certErrs...)
+
+	if numCerts := len(certs); numCerts != 1 {
+		return append(errs, fmt.Errorf("expected a single client certificate, got %d", numCerts))
+	}
+
+	cert := certs[0]
+	clientAuthEKUFound := false
+	for _, eku := range cert.ExtKeyUsage {
+		if eku == x509.ExtKeyUsageClientAuth {
+			clientAuthEKUFound = true
+			break
+		}
+	}
+	if !clientAuthEKUFound {
+		errs = append(errs, fmt.Errorf("expected the certificate to containe client authentication EKU"))
+	}
+
+	return errs
+}
+
+func validatePrivateKey(pemKey []byte) []error {
+	if len(pemKey) == 0 {
+		return []error{fmt.Errorf("required private key is empty")}
+	}
+	if _, err := keyutil.ParsePrivateKeyPEM(pemKey); err != nil {
+		return []error{fmt.Errorf("failed to parse the private key")}
```

Comment:
Would it make sense to wrap the original error to aid in troubleshooting?

Reply:
When it comes to private keys, I'd rather avoid printing errors that I don't directly control.